### ISSUE TITLE
Hyperion seems to skip an instruction when in trace mode after switch to z/Arch mode

### ISSUE THREAD

**Issue by arfineman** [SDL-Hercules-390/hyperion#295]
Same program works fine in 3.13.

Test Program: 
```
         LA    R1,1           
         SIGP  R0,R0,X'12'    
         BCTR  R1,R0          
         BCTR  R2,R0          <------- Instruction being skipped
         BCTR  R3,R0          
         LA    R1,0           
         SIGP  R0,R0,X'12'    
         SR    R15,R15        
         BR    R14          
```

Trace of the program:
```
-> 0000E000  LA      41100001  = 00000001    CC 1               
   0000E004  SIGP    AE000012    00000012    CC 0               
   000000000000E008   BCTR    0610        CC 0                  
-> 000000000000E00C   BCTR    0630        CC 0                  
   000000000000E00E   LA      41100000  = 00000000    CC 0      
   000000000000E012   SIGP    AE000012    00000012    CC 0      
   0000E016  SR      1BFF        CC 0                           
   0000E018  BR      07FE     -> 00000000    CC 0                 
```

**Comment by Fish-Git** [comment]
Need more information:

1. What version of Hercules are you running? What is the output of the Hercules `version` command?

2. What guest operating system are you running, and what type of "trace" are you doing? The trace output you provided does not appear to be coming from Hercules.

3. (Probably not needed but the more information the better) What host operating system are you using? Windows? Linux? Mac? Other?

4. Etc...

In short: you need to provide more detailing information regarding your perceived problem so that we can try to reproduce it for ourselves. If we cannot reproduce your problem then we cannot fix it. Before we can help you, you need to help us by providing enough detailed information for us to be able to look into it! &nbsp; _(See: **[How to Ask for Help](https://github.com/Fish-Git/AskForHelp)**)_

**Comment by arfineman** [comment]
Hi, Since you mentioned the release, I went back and tested several releases and it works on all of them. It only fails on 4.2.1.9826. So, that narrows the problem a lot.
Here is the result from 4.2.0.9708:
 -> 00020000  LA      41100001  = 00000001    CC 1          
    00020004  SIGP    AE000012    00000012    CC 0          
    0000000000020008   BCTR    0610        CC 0             
    000000000002000A   BCTR    0620        CC 0             
    000000000002000C   BCTR    0630        CC 0             
    000000000002000E   LA      41100000  = 00000000    CC 0 
    0000000000020012   SIGP    AE000012    00000012    CC 0 
    00020016  SR      1BFF        CC 0                      
    00020018  BR      07FE     -> 00F1294C    CC 0        
Here is how to reproduce it on any zVM system:
Assemble the test program above and then:
CP TR INST R 20000.FF RUN
LOAD TEST ( ORIGIN 20000 START

**Comment by Fish-Git** [comment]
Thank you! I will look into this and get back to you.

**Comment by Fish-Git** [comment]
Wow. Just...... _wow._ &nbsp; _&lt;me: shocked!&gt;_  &nbsp; :open_mouth: 

I was able to recreate the problem and you're absolutely right! With 4.2.0 it works fine, but with 4.2.1, it fails _exactly_ as you describe! (Wow!)

I am absolutely blown away.

_**Something is seriously wrong somewhere!** &nbsp;(to say the least!)_

_(sigh!)_

Let me start looking into it... &nbsp; `:(`

@ivan-w ? @wably ? You two know VM and PER way better than I do. I'll probably need your help on this one! I'll try to start digging into it myself, but I'm not real strong on Herc's SIE and VM and PER support, so I don't know how much (how well) I'll be able to do. I'll certainly _try_, but this might be beyond my ability to debug and would appreciate any help either of you _(or anyone really!)_ can provide. Thanks.

**Comment by arfineman** [comment]
I think at some point when the mode is switched and trace is active, the ILC (Instruction Length Counter) is erroneously advance by 2 extra bytes.

**Comment by wably** [comment]
I was able to recreate this as documented.  However, I did some additional tests and I believe this to be a SIE bug and not a PER bug.  I used the SATK tool to use this same example program modified slightly to set up PER instruction fetch events.  The modified program was then run natively on Hercules without an operating system present.  Upon the switch to z/Architecture mode, the SIGP instruction did not cause the BCTR R2,0 instruction to be skipped.

I've attached a small zip file here with the little program, the loadcore image, and the resulting Hercules log with Hercules tracing active.  You can see each instruction executed and the PER interruptions occurring.  As stated, the BCTR R2,0 is not skipped.

I think this eliminates PER as the cause but I will caution that conclusion with these caveats:

- I only enabled the instruction fetch (IF) event.  It could be that the CP TRACE INST command enables other event bits besides IF, but why would it?  Instructions are the only trace event being asked for.
- There is a small possibility that this could be a bug in z/VM and not Hercules. Maybe there is a small problem with the CP simulation of SIGP and interaction with PER in virtual machine.  Maybe someone who has a z/PDT or access to real iron can confirm if the original example works or fails as documented; it is so easy to cut/paste recreate the original example program above and turn on the CP TRACE.

Other than those two possible exceptions, this looks more like an SIE issue to me.  Ivan?

Regards,
Bob
[sigper.zip](https://github.com/SDL-Hercules-390/hyperion/files/4566157/sigper.zip)

**Comment by wably** [comment]
I just realized that it cannot be a bug in z/VM itself, since it was reported that earlier Hercules versions did not exhibit the problem, so please ignore that point.

Bob

**Comment by Fish-Git** [comment]
Bob,

I believe it might still be a PER bug, but one related to SIE (i.e. one that only occurs (is only triggered) when PER is enabled while running under SIE).

Whatever the bug is and whatever is causing it, the bug appears to be related to the instruction address in the PSW being improperly bumped (incremented) by an extra 2 bytes on the next instruction immediately after switching architecture modes:

```
/* TESTn */                      
                                 
'CP TR INST R 20000.FF RUN'      
'LOAD TESTn ( ORIGIN 20000 START'
'CP TR END ALL'                  

(where 'n' of course is 1, 2 or 3, etc)

--------------------------------------------------------------------------------
TEST2 ASSEMBLE:

TEST     CSECT             
R0       EQU   0           
R1       EQU   1           
R2       EQU   2           
R3       EQU   3           
R4       EQU   4           
R5       EQU   5           
R6       EQU   6           
R7       EQU   7           
R8       EQU   8           
R9       EQU   9           
R10      EQU   10          
R11      EQU   11          
R12      EQU   12          
R13      EQU   13          
R14      EQU   14          
R15      EQU   15          
*                          
         LA    R1,1        
         SIGP  R0,R0,X'12' 
*                          
         LA    R0,0        
         BCTR  R0,R0       
         LA    R1,1        
         BCTR  R1,R0       
         LA    R2,2        
         BCTR  R2,R0       
         LA    R3,3        
         BCTR  R3,R0       
*                          
         LA    R1,0        
         SIGP  R0,R0,X'12' 
*                          
         SR    R15,R15     
         BR    R14         
         END               
--------------------------------------------------------------------------------
test2                                                      
DMSLIO740I Execution begins...                             
 -> 00020000  LA      41100001  = 00000001    CC 1         
    00020004  SIGP    AE000012    00000012    CC 0         
    0000000000020008   LA      41000000  = 00000000    CC 0
 -> 000000000002000E   LA      41100001  = 00000001    CC 0
    0000000000020012   BCTR    0610        CC 0            
    0000000000020014   LA      41200002  = 00000002    CC 0
    0000000000020018   BCTR    0620        CC 0            
    000000000002001A   LA      41300003  = 00000003    CC 0
    000000000002001E   BCTR    0630        CC 0            
    0000000000020020   LA      41100000  = 00000000    CC 0
    0000000000020024   SIGP    AE000012    00000012    CC 0
    00020028  SR      1BFF        CC 0                     
    0002002A  BR      07FE     -> 00F0EAC4    CC 0         
Trace ended                                                
Ready; T=0.01/0.01 17:24:29                                
--------------------------------------------------------------------------------
TEST3 ASSEMBLE:

TEST     CSECT             
R0       EQU   0           
R1       EQU   1           
R2       EQU   2           
R3       EQU   3           
R4       EQU   4           
R5       EQU   5           
R6       EQU   6           
R7       EQU   7           
R8       EQU   8           
R9       EQU   9           
R10      EQU   10          
R11      EQU   11          
R12      EQU   12          
R13      EQU   13          
R14      EQU   14          
R15      EQU   15          
*                          
         LA    R1,1        
         SIGP  R0,R0,X'12' 
*                          
         LA    R0,0        
         LA    R1,1        
         LA    R2,2        
         LA    R3,3        
*                          
         LA    R1,0        
         SIGP  R0,R0,X'12' 
*                          
         SR    R15,R15     
         BR    R14         
         END               

--------------------------------------------------------------------------------
test3                                                       
DMSLIO740I Execution begins...                              
 -> 00020000  LA      41100001  = 00000001    CC 1          
    00020004  SIGP    AE000012    00000012    CC 0          
    0000000000020008   LA      41000000  = 00000000    CC 0 
 -> 000000000002000E   ????    0001                         
*** 000000000002000E       PROG    0001 -> 00000000         
              OPERATION                                     
HCPGIR453W CP entered; program interrupt loop               
--------------------------------------------------------------------------------
```

Interesting, yes?

**Comment by arfineman** [comment]
The good news is that the bug is confined to changes made between 4.2.0 and 4.2.1.

**Comment by Fish-Git** [comment]
> The good news is that the bug is confined to changes made between 4.2.0 and 4.2.1.

Well, I wouldn't exactly call it _good_ news, but yes, the bug, whatever it is, is definitely in 4.2.1++.

I'm on it! &nbsp; _(just as I hope Bob and/or Ivan and/or anyone/everyone else is too!)_

**Comment by ivan-w** [comment]
Once I have the test case set I'll just d a god ole git bisect !

**Comment by Fish-Git** [comment]
> Once I have the test case set I'll just do a god ole git bisect!

No need!

_**I FOUND IT!**_

It's commit 4998c658b38731fb7e74ff54aa66e9fa4521f0f8 that broke it. ("Workaround fix for poor z/VM z/Arch guest performance")

I don't fully understand _why (or how!)_ it broke it, but it did.

The commit immediately before that one works fine. But _that_ commit -- 4998c658b38731fb7e74ff54aa66e9fa4521f0f8 -- introduced the problem.

Still scratching my head... &nbsp; `:(`

**Comment by ivan-w** [comment]
Cool ! (but it actually raised another problem which ma require opening a separate issue)... I can't get git bisect to work because it always reports the latest git version (whatever source version I am using)

**Comment by Fish-Git** [comment]
Well I have good news and bad news.

The good news is I have a fix! &nbsp; `:)`

The bad news is it likely breaks something else. &nbsp; `:(`

Specifically, issue #222 _(although I admit to not having tested/verified that yet)._

The fix is to basically revert the temporary fix for Issue #222:

```diff
--- hyperion-1/sie.c	2019-08-24 19:38:33.964676400 -0700
+++ hyperion-0/sie.c	2020-05-03 09:49:19.238515900 -0700
@@ -325,14 +325,7 @@
 #if defined( FEATURE_001_ZARCH_INSTALLED_FACILITY )
     if (STATEBK->mx & SIE_MX_ESAME)
     {
-        /* If GUESTREGS was initialized for a different architecture,
-           then we need to initialize it for the new architecture */
-        if (GUESTREGS->arch_mode != ARCH_900_IDX)
-        {
-            /* REGS 'arch_mode' MUST be set BEFORE calling cpu_init! */
-            GUESTREGS->arch_mode = ARCH_900_IDX;
-            cpu_init( regs->cpuad, GUESTREGS, regs );
-        }
+        GUESTREGS->arch_mode = ARCH_900_IDX;
         GUESTREGS->program_interrupt = &z900_program_interrupt;
         icode = z900_load_psw( GUESTREGS, STATEBK->psw );
     }
```

The above patch puts the code in sie.c back to the way it originally was, before my attempt to fix #222. With the above patch in place, all tests now pass. arfineman's original test code behaves as expected, as do my own test cases too.

Additionally, reversing the architecture modes of the SIGPs in each test (so that each test switches to _S390_ mode first, executes a few instructions, and then switches back to the "original" z/Architecture mode) and running the tests in zcms mode (i.e. `ipl zcms`), all of the tests behave as expected as well.

That is to say, starting in normal cms mode and having each test switch to zarch, execute a few instructions, and then switch back to 390, _-OR-_ starting each test from zcms (zarch) mode and having each test switch to 390 mode, execute a few instructions and then switch back to zarch mode, they all behave as expected as well.

Without the above patch, they fail (some catastrophically).

So the above patch, applied to current git, _definitely_ fixes the problem. _BUT..._ as mentioned, it _might_ also break issue #222. So depending on the results of further issue #222 testing, if we apply the above patch as the fix for this issue, we _might_ also have to re-open issue #222 as a result.

Am beginning my #222 tests now... Will post results once I have them.

Thanks.

@ivan-w ? should we go ahead and apply the above patch anyway to fix this immediate tracing problem? I personally think we should. I think having correct tracing behavior is more important than a performance issue! What you do you think? Do you agree?

**Comment by Fish-Git** [comment]
> Am beginning my #222 tests now... Will post results once I have them.

Great news! #222 is still resolved!

That is to say, the above patch -- which fixes _**this**_ issue _(instr skip during trace)_ -- does _NOT_ re-break (regress) issue #222 (the speed fix)! Both `toto exec` and `rexxcps exec` were just as fast with the above patch applied as they were without the above patch applied.

So I think we're good! My fix is good! I'm going to go ahead and commit my fix and then close this issue as resolved.

Just as a sidebar, I believe the change that the above patch makes (that fixes the problem) probably _should_ have been made way back when my **["more proper fix" commit](https://github.com/SDL-Hercules-390/hyperion/commit/742a2d1065faa4e733efc76922176d54344e4921)** was made, but I simply forgot to do it at the time.

**Comment by Fish-Git** [comment]
Fixed by commit f5504fe40769a3ad666d85631db68302a44c2ee2.

@arfineman ? @ivan-w ? Can you please confirm? Thanks!

Regardless, am closing this issue as resolved.

**Comment by arfineman** [comment]
Thank you. Is there a way to test this fix without having to re-build Hercules ? If that's the case, I'll have to wait until the next build is released. Best regards,

**Comment by Fish-Git** [comment]
> Is there a way to test this fix without having to re-build Hercules?

Unfortunately, no. The fix was committed to the repository, so to get the fix, you will need to use a version of Hercules that was built _WITH_ the fix already in it, which means you will have to either clone the current version of the our repository (that has the fix) and build it, or else wait for 4.3 to be officially released to get the prebuilt with the fix already in it.

**Comment by ivan-w** [comment]
Confirmed and confirmed !

**Comment by ivan-w** [comment]
```
i zcms                         
z/CMS V6.3.0   2018-08-17 16:08
DMSACP723I T (592) R/O         
Ready; T=0.08/0.11 04:38:12    
rexxcps 1000                                            
----- REXXCPS 2.2 -- Measuring REXX clauses/second -----
 REXX version is: REXX370 4.02 01 Dec 1998              
       System is: CMS                                   
       Averaging: 1000 measures of 30 iterations        
                                                        
     Performance: 70460 REXX clauses per second         
                                                        
Ready; T=419.35/420.89 04:05:30                         
i cms                                                   
z/VM V6.3.0    2018-08-17 16:08                         
                                                        
DMSACP723I T (592) R/O                                  
Ready; T=0.05/0.06 04:08:29                             
rexxcps 1000                                            
----- REXXCPS 2.2 -- Measuring REXX clauses/second -----
 REXX version is: REXX370 4.02 01 Dec 1998              
       System is: CMS                                   
       Averaging: 1000 measures of 30 iterations        
                                                        
     Performance: 71723 REXX clauses per second         
                                                        
Ready; T=411.78/413.35 04:15:31                         
```
```
tstskip                                                    
 -> 00014000  BALR    05C0        CC 0                     
GPR  0 =  00000E08  0B000848  00000848  80F0E024           
GPR  4 =  00002D78  00F0D980  00F0E980  00F0FEA8           
GPR  8 =  00000000  E3E2E3E2  D2C9D740  00F0E52A           
GPR 12 =  40014002  0000A668  00F0EAC4  00014000           
    00014002  LA      41100001  = 00000001    CC 0         
GPR  0 =  00000E08  00000001  00000848  80F0E024           
GPR  4 =  00002D78  00F0D980  00F0E980  00F0FEA8           
GPR  8 =  00000000  E3E2E3E2  D2C9D740  00F0E52A           
GPR 12 =  40014002  0000A668  00F0EAC4  00014000           
    00014006  LA      4180000A  = 0000000A    CC 0         
GPR  0 =  00000E08  00000001  00000848  80F0E024           
GPR  4 =  00002D78  00F0D980  00F0E980  00F0FEA8           
GPR  8 =  0000000A  E3E2E3E2  D2C9D740  00F0E52A           
GPR 12 =  40014002  0000A668  00F0EAC4  00014000           
    0001400A  SIGP    AE000012    00000012    CC 0         
GRG  0 =  0000000000000E08  0000000000000001               
GRG  2 =  0000000000000848  0000000080F0E024               
GRG  4 =  0000000000002D78  0000000000F0D980               
GRG  6 =  0000000000F0E980  0000000000F0FEA8               
GRG  8 =  000000000000000A  00000000E3E2E3E2               
GRG 10 =  00000000D2C9D740  0000000000F0E52A               
GRG 12 =  0000000040014002  000000000000A668               
GRG 14 =  0000000000F0EAC4  0000000000014000               
    000000000001400E   BCTR    0610        CC 0            
GRG  0 =  0000000000000E08  0000000000000000               
GRG  2 =  0000000000000848  0000000080F0E024               
GRG  4 =  0000000000002D78  0000000000F0D980               
GRG  6 =  0000000000F0E980  0000000000F0FEA8               
GRG  8 =  000000000000000A  00000000E3E2E3E2               
GRG 10 =  00000000D2C9D740  0000000000F0E52A               
GRG 12 =  0000000040014002  000000000000A668               
GRG 14 =  0000000000F0EAC4  0000000000014000               
    0000000000014010   BCTR    0680        CC 0            
GRG  0 =  0000000000000E08  0000000000000000               
GRG  2 =  0000000000000848  0000000080F0E024               
GRG  4 =  0000000000002D78  0000000000F0D980               
GRG  6 =  0000000000F0E980  0000000000F0FEA8               
GRG  8 =  0000000000000009  00000000E3E2E3E2               
GRG 10 =  00000000D2C9D740  0000000000F0E52A               
GRG 12 =  0000000040014002  000000000000A668               
GRG 14 =  0000000000F0EAC4  0000000000014000               
    0000000000014012   BCTR    0630        CC 0            
GRG  0 =  0000000000000E08  0000000000000000               
GRG  2 =  0000000000000848  0000000080F0E023               
GRG  4 =  0000000000002D78  0000000000F0D980               
GRG  6 =  0000000000F0E980  0000000000F0FEA8               
GRG  8 =  0000000000000009  00000000E3E2E3E2               
GRG 10 =  00000000D2C9D740  0000000000F0E52A               
GRG 12 =  0000000040014002  000000000000A668               
GRG 14 =  0000000000F0EAC4  0000000000014000               
    0000000000014014   LA      41100000  = 00000000    CC 0
GRG  0 =  0000000000000E08  0000000000000000 
GRG  2 =  0000000000000848  0000000080F0E023               
GRG  4 =  0000000000002D78  0000000000F0D980               
GRG  6 =  0000000000F0E980  0000000000F0FEA8               
GRG  8 =  0000000000000009  00000000E3E2E3E2               
GRG 10 =  00000000D2C9D740  0000000000F0E52A               
GRG 12 =  0000000040014002  000000000000A668               
GRG 14 =  0000000000F0EAC4  0000000000014000               
    0000000000014018   SIGP    AE000012    00000012    CC 0
GPR  0 =  00000E08  00000000  00000848  80F0E023           
GPR  4 =  00002D78  00F0D980  00F0E980  00F0FEA8           
GPR  8 =  00000009  E3E2E3E2  D2C9D740  00F0E52A           
GPR 12 =  40014002  0000A668  00F0EAC4  00014000           
    0001401C  LR      18F8        CC 0                     
GPR  0 =  00000E08  00000000  00000848  80F0E023           
GPR  4 =  00002D78  00F0D980  00F0E980  00F0FEA8           
GPR  8 =  00000009  E3E2E3E2  D2C9D740  00F0E52A           
GPR 12 =  40014002  0000A668  00F0EAC4  00000009           
    0001401E  BR      07FE     -> 00F0EAC4    CC 0         
GPR  0 =  00000E08  00000000  00000848  80F0E023           
GPR  4 =  00002D78  00F0D980  00F0E980  00F0FEA8           
GPR  8 =  00000009  E3E2E3E2  D2C9D740  00F0E52A           
GPR 12 =  40014002  0000A668  00F0EAC4  00000009           
Ready(00009); T=0.01/0.02 04:41:10                         
```
